lang: add the `.tailcall` calling convention

zerbina opened the pull request:
## Summary

Add the `.tailcall` calling convention to the language. It designates
routines as being guaranteed to take over their caller's stack frame,
but only when the caller also uses the `.tailcall` calling convention.

## Details

### Design

Guaranteed tail-call elimination is an important requirement for
programs written in functional style, continuation-passing styles, or
any other style that requires deep, potentially unbounded recursion.

Implementing it efficiently via a macro is hardly possible, if at all,
making it a good candidate for being a built-in language feature.

Only enforcing tail calls for `.tailcall` calls within `.tailcall`
routines, removes the need for adapter routines at the non-`.tailcall`
to `.tailcall` edge, but it also makes the static semantics of
`.tailcall` calls callsite-dependent.

### Implementation

A description of how the portable tail-call elimination works is
provided in the `tailcallelim.rst`.

The implementation is scattered throughout multiple stages:
* `sem` and `sempass2` take care of the analysis part and synthesizing/
  instantiating the needed types and routines
* `transf` and `mirgen` perform the portable tail call elimination
* `injecthooks` handles the "trailing destructor call" detection and
  related reporting
* the code generators implement storing the parameter tuple into the
  type-erased storage (the new `mStoreParams` magic)

Whether `mirgen` performs the portable tail call elimination is
configured through the `goTailCallElim` option. When not present,
must-become-sibling-call calls are translated to the new `mnkTailCall`
MIR operation.

The VM backend uses true sibling calls, as the `mStoreParams` magic
cannot be implemented by the VM due to its strict typing requirements.

zerbina committed:
manual: document `.musttail` and tail-call semantics

zerbina committed:
tests: cover the semantics laid out by the manual

zerbina committed:
document tail call elimination approach

zerbina committed:
disallow `.musttail` for methods

Methods using the `.musttail` calling convention is too much effort to
make work, but it's not fundamentally impossible.

zerbina committed:
register the `.musttail` calling convention

zerbina committed:
wordrecg: register `musttail` as word

zerbina committed:
sem: instantiate `Continuation` type for `.musttail` proc

Sem is the only layer where generic types can be properly instantiate, and
therefore the return `Continuation` types are instantiated right when a
`.musttail` procedure type is constructed.

The type must not be visible to macros and the rest of sem, so it's hidden
away in a new effect-list slot of the `tyProc` type. `sempass2` is changed
such that it doesn't clobber the slot. Ultimately, this is a massive hack.

zerbina committed:
sempass2: implement tail call analysis

It verifies that all calls to `.musttail` procedure adhere to the
rules laid out by the manual.

Synthesizing the apply and trampoline procedures also happens in
`sempass2` already, in order to side-step issues with IC and too-late
generic instantiations.

zerbina committed:
sem: wrap routine expression bodies in `return`

This removes the need to add any special cases in order to support,
e.g.: `proc a(): int = tail()`. A standalone yet trailing
`result = tail()` is not accepted as a tail call.

Visible to macros and thus - in theory - a breaking change.

zerbina committed:
transf: implement tail-call elimination

zerbina committed:
transf: insert `return` after void `.musttail` call

This makes the tail-call elimination for void calls a lot easier.

zerbina committed:
transf: implement `return` forwarding

It's easier to implement tail-call elimination this way, since tail
calls are guaranteed to appear directly as the `return`'s operand.

To keep compile times in check, the pass is only enabled for tail-call-
elimination-pass-using procedures.

zerbina committed:
mirgen: enforce borrows for by-ref `.musttail` arguments

Normal argument handling might introduce intermediate temporaries,
which is unacceptable for `.musttail` calls, as these procedure rely
on all by-ref locations outliving the call.

zerbina committed:
mirgen, cgen: handle `Continuation` return types

The real return types and hidden environment parameter are not visible
on the `tyProc` types, requiring special handling by both `mirgen` and
`cgen`.

`mirgen` registers uses the real return type for the result variable,
and `cgen` looks at the MIR body instead of the `PType` for the return
type. The hidden env parameter is also special case by `cgen`.

zerbina committed:
mirgen: turn the old result variable into a local

zerbina committed:
mirtypes: lower `.musttail` types right away

This is necessary because the `Continuation` to use for the type
parameter is not accessible when doing proper type lowering (in
`lowerType`).

zerbina committed:
mirgen: patch `.musttail` proc symbols

Both `.musttail` procs and the trampoline-requiring procs are replaced
with the apply and trampoline procedures, respectively, but only
outside of generated ops.

zerbina committed:
mirgen: handle the new magics

zerbina committed:
fix code generation for calls to patched procedures

Their callee type is wrong (the return type is not patched), and thus
the correct expression type has to be taken from the call itself.

This is changed in both cgen and mirgen. The other code generators
probably need the fix too.

zerbina committed:
cgen: implement `mStoreParams` magic

zerbina committed:
types: always treat array as pass-by-reference

They're always passed by reference when using the C backend, therefore
they also need to be considered as such by tail call elimination.

alaviss commented:
Is it feasible to add a helper/pragma to allow calling `musttail` procs outside of tail positions (i.e. by not forcing a transform of the caller)?

It's trivial enough to write a quick proc wrapper around such calls, but it'd be better to just tell the compiler not to transform the caller.

zerbina commented:
> Is it feasible to add a helper/pragma to allow calling `musttail` procs outside of tail positions (i.e. by not forcing a transform of the caller)?

An expression pragma forcing the `.musttail` call to be trampolined in-place is doable with a reasonable amount of effort, yes.

zerbina committed:
docs: link to `tailcallelim.rst`

zerbina committed:
manual: clarify what a tail call is

zerbina committed:
sem: fix crash with tail-call analysis in doc mode

The untyped AST in `runnableExample` sections was not accounted for...

zerbina committed:
sem: properly handle `except` clauses in tail-call analysis

Their content cannot be the last expression/statement in the presence
of a `finally` clause.

zerbina committed:
manual: adjust the tailing `return` rules

saem commented on compiler/sem/semtypes.nim:
```diff
@@ -1928,6 +1945,10 @@ proc semProcTypeWithScope(c: PContext, n: PNode,
     # we're still interested in implicit tags and raises pragmas
     n[1] = implicitPragmas(c, s, n[1], {wTags, wRaises})
 
+  # instantiat the type of the continuation for .musttail procedures
```
```suggestion
  # instantiate the type of the continuation for .musttail procedures
```

saem commented on doc/manual.rst:
```diff
@@ -4101,6 +4107,59 @@ Overloading of the subscript operator
+A call is considered a tail call iff:
+* it is the *tailing expression* of a `return`, and the `return` is a
+  *tailing return*
+* it is the *tailing expression* of a rotine's body
```
```suggestion
* it is the *tailing expression* of a routine's body
```

saem commented on doc/manual.rst:
```diff
@@ -4101,6 +4107,59 @@ Overloading of the subscript operator
+----------------
+
+Since `.musttail` routine calls are always they take over their caller's stack
+frame, some restriction are placed on their arguments:
```
```suggestion
Since `.musttail` routine calls always take over their caller's stack frame,
some restrictions are placed on their arguments:
```

saem commented on doc/manual.rst:
```diff
@@ -4101,6 +4107,59 @@ Overloading of the subscript operator
+* arguments to `var`, pass-by-reference, or pass-by-value-with-overridden-copy-
+  operator type parameters must be lvalue expressions derived from caller
+  parameters or globals
+* pass-by-value parameters without for whose type `supportsCopyMem` returns
+  `true`` act like `sink` parameters
```
```suggestion
* pass-by-value parameters with a type for which `supportsCopyMem` returns
  `true` act like `sink` parameters
```

saem reviewed (commented):
partial review, I still have to go through the tests properly.

---

Current Thoughts

I like the change, my only suggestion is that the calling convention should perhaps be called `tailcall`, my reasoning is as follows:
* the calling conventions are pretty much all things one must do/adhere to anyways, so that's redundant
* `tailcall` is easier to search both in the doc (fewer ambiguities) and on the internet, for the general concept
* `musttail` is a bit overloaded term with people who have a gcc/llvm background making all sorts of misplaced assumptions

zerbina committed:
manual: reorder the "trailing return" listing

zerbina committed:
sem: skip `runnableExamples` entirely

Their argument block is untyped, which only leads to crashes/errors
when they're processed. They're now skipped entirely.

zerbina committed:
sem: fix compiler crash

How hard can making the analysis work be, you may ask?

zerbina committed:
sem: special-case `for` in tail call analysis

zerbina committed:
jsgen: implement `mStoreParams` magic

The implementation is super hack-y and relies heavily on the callsite
env argument having the proper shape (in other words, it's brittle).
Not ideal, but at least it works.

zerbina committed:
jsgen: handle the `.musttail` hidden env parameter

zerbina committed:
tailcall_elim: work around `jsgen` bugs

*Converting* from `ptr T` to `pointer` doesn't work (it's not
implemented correctly), but *casting* does, so use a cast.

Of course, it'd be better to implement conversion code generation
properly instead.

zerbina committed:
jsgen: handle continuation return types correctly

zerbina committed:
backends: don't create empty CGIR bodies

They violate the "result slot always exists" violation. The workaround
in `cgen` is removed again.

zerbina committed:
tests: mark `openArray` test as known issue

`toOpenArray` is not yet fully supported by the JS code generator.

saem commented on compiler/sem/tailcall_elim.nim:
```diff
@@ -0,0 +1,525 @@
+# -------------- analysis --------------
+
+proc checkArg(g: ModuleGraph, owner: PSym, n: PNode, i: int, formal: PType) =
+  ## Analyses the `i`-th argument `n` for an call to a procedure with type
```
```suggestion
  ## Analyses the `i`-th argument `n` for a call to a procedure with type
```

saem commented on compiler/sem/tailcall_elim.nim:
```diff
@@ -0,0 +1,525 @@
+  # signature cannot easily and transparently be changed to something else
+  # cannot use tail calls
+  if s.typ == nil:
+    echo s
```
Leftover debug code?

zerbina replied:
Yep, thanks, I'd have likely missed that.

saem commented on compiler/sem/tailcall_elim.nim:
```diff
@@ -0,0 +1,525 @@
+      recurse(n[0][1], mode + {emLast} - {emExpr})
+  of nkCallKinds:
+    if n[0].kind == nkSym and n[0].sym.magic == mRunnableExamples:
+      # the body was not type properly, nor is it relevant for the
```
```suggestion
      # the body was not typed properly, nor is it relevant for the
```

saem reviewed (approved):
Went through the rest of the code as is, I didn't go quite as a deep as I know there is a direction change to remove the caller transform requirement, along with associated restrictions (around iterator, exportc, methods, etc), coming.

I can do another pass after that, if you'd like.

zerbina committed:
new semantics (step 1)

Update the manual with the new semantics of `.musttail` calls.

zerbina committed:
new semantics (step 3)

Only check `.musttail` calls when in `.musttail` context.

zerbina committed:
new semantics (step 2)

Remove the `sfCallsMusttail` flag and everything directly related to
it.

zerbina committed:
manual: remove the erroneous `sink` clause

Copy-to-`sink` is very much allowed for tail calls.

zerbina committed:
manual: no tail calls in all `except`/`finally` clauses

Every `except` clause performs implicit cleanup at its end (to pop the
handled exception from the stack), therefore no tail call is possible
there.

For `finally` clauses, where they exit to cannot be known statically,
therefore the next statement/expression cannot be known statically and
thus no tail call is possible there.

zerbina committed:
new semantics (step 4)

Implement the trampolining of `.musttail` calls in `mirgen`. This only
works for static calls, dynamic calls need dedicated handling.

zerbina committed:
move tail-call elimination into `mirgen`

It's not an ideal solution (a MIR pass would be better), but it is at
least better than the `transf`-based transformation, as the `mirgen`-
based lowering allows disabling it via a code generator option, which
is going to be necessary for VM support.

Some bugs with the previous implementation are fixed, too:
* the user-visible `result` variable in a `.musttail` procedure now
  behaves exactly like it does elsewhere
* an indirect `.musttail` call now works outside of a `.musttail`
  context

zerbina committed:
report an error for trailing cleanup

`injecthooks` is the most natural place to place the detection and
marker removal, because:
* it already has the facilities to emit errors
* it takes place after `destroy` elimination
* it's only run when destructor injection is enabled (which is the only
  situation where `ensureNoCleanup` markers may be present)

That said, `injecthooks` is still not a good fit, but a separate module
would be too much and `mirpasses` lacks the reporting bits.

zerbina committed:
transf: wrap all tail calls in `return`

All tail calls now use the form `nkReturn(nkCall(...))` in transformed
AST, making trivial to detect.

This breaks the current tail-call elimination pass, but it'll make
the MIR-based transformation a lot easier to implement.

zerbina committed:
mir: add the `mnkTailCall` operation

It's used for encoding where native tail (well, sibling) calls need to
take place.

zerbina committed:
mirgen: handle untyped `mMove` magics

The now obsolete `mEnsureNoMove` and `mStoreParams` magic handling is
removed, too.

zerbina committed:
sempass2: ensure `Continuation` types have hooks

zerbina committed:
vm: add the `opcTailCall` instruction

zerbina committed:
vmgen: emit `opcTailCall` for `cnkTailCall`

zerbina committed:
mirgen: make tail-call elimination togglable

Add an option for toggling tail-call elimination in `mirgen` on
and off.

It's enabled for both the C and JS backend, but not for the VM backend,
as the latter supports native tail calls.

When the option is not enabled, no tail-call elimination is performed
in `mirgen`; eligible tail calls are only translated to `mnkTailCall`
operations.

zerbina committed:
mirgen: partially remove the `tailExit` restrictions

When tail-call elimination is disabled for `mirgen`, the `result`
variable cleanup logic is emitted, which triggers the `unreachable`
in `tailExit` even though known to be harmless.

For convenience of `mirgen`, `bkTryExcept` blocks are allowed, under
the assumption that the callsite knows that they're harmless.

zerbina committed:
tests: mark `t02_params_openarray` as know issue

The VM doesn't support `toOpenArray`.

zerbina committed:
tests: update the tests

zerbina committed:
docgen: adjust to the new body transformation

Routines bodies that are expressions are now transformed into
`return expr` instead of `result = expr`. `docgen` had special handling
to undo the `result = expr` transformation (of course, what else),
which needs to be adjusted now.

zerbina committed:
cgir: remove `emptyBody` again

It only led to issues. Instead of passing an empty body to the
`startProc` call, `jsbackend` and `cbackend` now pass the initial body
to it.

zerbina committed:
improve some formatting

zerbina committed:
manual: `block` bodies are tailing expressions, too

zerbina committed:
tailcall_elim: better analysis + proper reporting

* use proper reports for errors
* use an analysis not using the flags
* fix some AST not being traversed

zerbina committed:
tests: update the `finally` tests

It still assumed the outdated rules.

zerbina committed:
rename `tailcall_elim` to `tailcall_analysis`

The module has only little to do with tail-call elimination anymore.

zerbina committed:
semexprs: don't wrap comments in `nkReturn`

Apply the return statement to the last expression in the list. This
also removes the need for the workarounds in `docgen` and `suggest`,
both which are removed.

zerbina committed:
sem: fix tail call analysis yet again

Will it be the final fix this time around? Tune in again next time, to
find out!

zerbina committed:
tests: update `tmacrostmt.nim`

The typed AST for routines with an expression body changed.

zerbina committed:
update `tailcallelim.rst`

* update the "invoke from non-musttail routine" section to reflect
  reality
* change the field names of `Continuation` to those used in the actual
  implementation
* improve some phrasing

zerbina committed:
rename "musttail" to "tailcall"

This includes the calling convention itself as well as all occurrences
in comments, reports, documentation, and other names.

Only the test director is not renamed yet.

zerbina committed:
sem: report error for unsized or too large parameters

zerbina committed:
rename the test directory

zerbina committed:
manual: update the `tailcall` documentation

It still described the old semantics.

zerbina committed:
fix some typos

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

zerbina committed:
Merge branch 'devel' into musttail-calling-convention

zerbina committed:
allow borrows from pointer dereferences

Pointers are unchecked and untracked in general; disallowing using a
pointer dereference (or a projection thereof) provides no real safety,
while preventing valid use cases.

zerbina committed:
various small documentation/comment fixes

zerbina committed:
mangling: update the mangled name for `.taillcal`

Not that it matters; no `.tailcall` types exists in the MIR type
IR.

zerbina committed:
Merge branch 'devel' into musttail-calling-convention

saem commented on compiler/backend/ccgtypes.nim:
```diff
@@ -342,8 +342,8 @@ proc prepareParameters(m: BModule, t: PType): seq[TLoc] =
   if prc.typ.callConv == ccTailcall:
-    # add a valid enough loc that the header can be generated properly. The
-    # real loc is filled in later
+    # add a param loc that's valid enough for the header can be generated
```
```suggestion
    # add a param loc that's valid enough for the header to be generated
```

zerbina committed:
sem: set size of `openArray` of `var openArray`

They're treated as a struct of two pointers by `sizealignoffsetimpl`,
allowing them to be used in a tail-call parameter tuple.

zerbina committed:
tests: fix invalid test specs

zerbina committed:
tests: mark `t02_params_sink` as known issue

`vmgen` doesn't emit a register move (because those don't exist...) for
`sink` parameters, leading to the source's storage being deallocated
when it's a local.

zerbina committed:
fix typo

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

zerbina committed:
tailcall_analysis: fix incompatible type error

Use `pointer` as the parameter type for `apply` procedures, so that
their signature matches that expected by `Continuation.next` (i.e.,
`proc(p: pointer): Continuation[T]`).

alaviss commented on compiler/backend/cgirgen.nim:
```diff
@@ -388,7 +388,11 @@ proc callToIr(tree: MirBody, cl: var TranslateCl, n: MirNode,
+                    of mnkCheckedCall: cnkCheckedCall
+                    of mnkTailCall:    cnkTailCall
+                    else: unreachable()),
                    info, cl.map(n.typ))
```
Purely cosmetic, but I think it's worth moving this `case` into it's own `let callKind`.

alaviss commented on compiler/backend/jsbackend.nim:
```diff
@@ -122,7 +123,8 @@ proc generateCode*(graph: ModuleGraph, mlist: sink ModuleList) =
     globals = newGlobals(graph)
-    bconf = BackendConfig(tconfig: TranslationConfig(magicsToKeep: NonMagics))
+    bconf = BackendConfig(tconfig: TranslationConfig(magicsToKeep: NonMagics,
+                                                     options: {goTailCallElim}))
```
Once the inner expression requires breaking lines, I believe the outer should be broken down as well. It keeps the door open for adding things in the future.

```suggestion
    bconf = BackendConfig(
      tconfig: TranslationConfig(
        magicsToKeep: NonMagics,
        options: {goTailCallElim}
      )
    )
```

alaviss commented on compiler/front/cli_reporter.nim:
```diff
@@ -496,6 +496,14 @@ proc reportBody*(conf: ConfigRef, r: SemReport): string =
+               "a return"
+
+    of rsemArgumentMustBorrowFromParameter:
+      result = "argument doesn't borrow from parameter, global, or ptr " &
```
```suggestion
      result = "argument doesn't borrow from parameter, global, or pointer " &
```

I'd say to keep short forms out of diagnostics.

zerbina replied:
I agree, but the usage of `ptr` here was to express that this only applies to `ptr T` parameters, and not `pointer` parameters.

Having said that, borrowing from a `pointer` projection is not possible anyway (you'd have to cast it to `ptr` first), so I think using `pointer` here is fine.

alaviss commented on compiler/front/cli_reporter.nim:
```diff
@@ -1364,6 +1372,30 @@ proc reportBody*(conf: ConfigRef, r: SemReport): string =
       result = "an exception can only be re-raised within the scope of an" &
                " except, with no finally in-between"
 
+    of rsemCleanupPreventsTailCall:
+      result = "'$1' requires cleanup that prevents a tail call" %
+               [r.ast.render]
```
Nothing actionable now, but this is kinda hard to grasp for people without a firm understanding of the language.

Maybe we should create diagnostic explainer manual pages like Rust in the future.

alaviss commented on lib/system.nim:
```diff
@@ -3064,3 +3064,17 @@ when defined(nimDebugUtils):
+    ## tail-call elimination.
+    case done: bool
+    of false:
+      next: proc(env: pointer): Continuation[T] {.nimcall.}
```
Can we enforce that `next` cannot be `nil` and use it as the discriminator instead of `done`? That'd prevent us from wasting 7 bytes of alignment for `done`.

This doesn't have to be solved now fwiw.

zerbina replied:
Sure, that'd also shorten the construction logic.

zerbina replied:
Hm, on second thought, this wouldn't provide much of a benefit. Only `Continuation[void]` is reduced in size, and for other cases (such as when `T` is a type that cannot be zero-initialized), it would actually hurt efficiency instead, since `result` is then always default-initialized, regardless of whether `next` is `nil` or not.

alaviss replied:
> since `result` is then always default-initialized, regardless of whether `next` is `nil` or not.

Why is a `bool` discriminator different for this case?

zerbina replied:
Because it's a real discriminator. If you use:
```nim
type Continuation[T] = object
  next: ...
  result: T
```
then `next` and `result` are just two adjacent fields with no further relationship between them.

A `proc` value cannot be used as a discriminator, not even internally.

alaviss commented on tests/lang/s99_tailcall/t02_params_by_ref.nim:
Quick question: should there be tests for `{.byref.}` and `{.bycopy.}` pragmas as well?

saem replied:
`byref` and `bycopy` are questionable features, so not bothering with proliferating testing for them seems reasonable.

alaviss reviewed (approved):
Only minor styling points and some questions, none of which are blocking.

Looks great otherwise to me.

alaviss commented:
Minor but non-blocking issue with diagnostics:

```nim
import macros

proc fib(c: int, a = 0, b = 1): int {.tailcall.} =
  case c
  of 0: a
  of 1: b
  else: fib(c - 1, b, a + b)

const BigFib = fib(50)

proc main() =
  echo fib()

main()
```

yields:

```
test.nim(12, 11) Error: type mismatch: got <>
but expected one of:
proc fib[type node](c: int; a = 0; b = 1; :env): int

expression: fib()
```

The internal implementation leaked into errors.

zerbina commented:
> Minor but non-blocking issue with diagnostics:

This is hard to fix properly, as `transf` currently needs to modify symbols in-place (`.closure`s are affected by the same problem), but it can be worked/hacked around with by modifying `renderer`.

**Edit:** I've implemented the workaround.

zerbina committed:
renderer: don't show hidden parameters

Could be made configurable, but wanting to have the env parameter shown
in `renderTree` output is such a niche use case that it's not worth the
effort.

zerbina committed:
cosmetic changes; error message adjustments

Co-authored-by: leorize+oss@disroot.org

zerbina commented:
/merge

Thanks for the reviews, everyone!

saem commented:
/merge
